### PR TITLE
GH-41317: [C++] Fix crash on invalid Parquet file

### PR DESCRIPTION
**Pull request by rouault**
### Rationale for this change

Fixes the crash detailed in #41317 in TableBatchReader::ReadNext() on a corrupted Parquet file

### What changes are included in this PR?

Add a validation that all read columns have the same size

### Are these changes tested?

I've tested on the reproducer I provided in #41317 that it now triggers a clean error:
```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    [_ for _ in parquet_file.iter_batches()]
  File "test.py", line 3, in <listcomp>
    [_ for _ in parquet_file.iter_batches()]
  File "pyarrow/_parquet.pyx", line 1587, in iter_batches
  File "pyarrow/error.pxi", line 91, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: columns do not have the same size
```
I'm not sure if/how unit tests for corrupted datasets should be added

### Are there any user-facing changes?

No

**This PR contains a "Critical Fix".**

* GitHub Issue: #41317

### CODE REVIEW

**Review by mapleFU — APPROVED** [review]

**Comment by mapleFU** [comment]
Will merge in 2days if no negative comments

**Review by felipecrv — CHANGES_REQUESTED** [review]
Can you add the `DCHECK` in the `TableBatchReader` as well?

And note in the class docstring that the `Table` is expected to be valid prior to using it with the batch reader?

**Comment by rouault** [comment]
> is expected to be valid prior to using it with the batch reader

done

**Comment by rouault** [comment]
It seems the added ValidateFull() breaks one test
```
[----------] 1 test from TestTableSortIndicesForTemporal/1, where TypeParam = class arrow::Date64Type
[ RUN      ] TestTableSortIndicesForTemporal/1.NoNull
D:/a/arrow/arrow/cpp/src/arrow/table.cc:622:  Check failed: _s.ok() Operation failed: table_.ValidateFull()
Bad status: Invalid: Column 0: In chunk 0: Invalid: date64[ms] 1 does not represent a whole number of days
```
I will have to let more knownledgeable people of the code base than me to investigate that.

File: cpp/src/arrow/table.cc [hunk 1]
```diff
@@ -619,6 +619,7 @@ TableBatchReader::TableBatchReader(const Table& table)
   for (int i = 0; i < table.num_columns(); ++i) {
     column_data_[i] = table.column(i).get();
   }
+  ARROW_DCHECK_OK(table_.ValidateFull());
```
Comment (felipecrv) [inline cpp/src/arrow/table.cc:622]: Please use `DCHECK(table_.Validate().ok());` so it's not validated in all builds.

`DCHECK_OK` is weird
https://github.com/apache/arrow/blob/192de026ad5c021846d89449757b0ac679436d13/cpp/src/arrow/util/logging.h#L92

File: cpp/src/arrow/table.cc [hunk 2]
```diff
@@ -632,6 +633,7 @@ TableBatchReader::TableBatchReader(std::shared_ptr<Table> table)
   for (int i = 0; i < owned_table_->num_columns(); ++i) {
     column_data_[i] = owned_table_->column(i).get();
   }
+  ARROW_DCHECK_OK(table_.ValidateFull());
```
Comment (felipecrv) [inline cpp/src/arrow/table.cc:636]: same

**Review by felipecrv — COMMENTED** [review]
If you use `Validate()` instead of `ValidateFull()` this won't be a problem.

`Validate()` contains the structural checks that ensure memory safety whereas `ValidateFull()` can get very deep into the validity of individual values of an array and is very slow.

https://github.com/apache/arrow/blob/192de026ad5c021846d89449757b0ac679436d13/cpp/src/arrow/array/validate.cc#L189-L193

File: cpp/src/arrow/table.cc (hunk 1 above)
Comment (rouault) [reply cpp/src/arrow/table.cc:622]: > Please use `DCHECK(table_.Validate().ok());` so it's not validated in all builds.

thanks for the suggestion. Commit amended with it.

**Comment by rouault** [comment]
This PR should be ready for merge (as far as I can see the test failures are also found in other pull requests)

File: cpp/src/arrow/table.h [hunk 3]
```diff
@@ -241,6 +241,9 @@ class ARROW_EXPORT Table {
 ///
 /// The conversion is zero-copy: each record batch is a view over a slice
 /// of the table's columns.
+///
+/// The table is expected to be valid prior to using it with the batch reader.
+///
```
Comment (mapleFU) [inline cpp/src/arrow/table.h:246]: nit: you can remove this line here

**Review by mapleFU — APPROVED** [review]
Will wait @felipecrv comment here

**Review by felipecrv — APPROVED** [review]

**Comment by mapleFU** [comment]
Will merge if no negative comment in one day

**Pull request merged**